Fix: assertion failure caused by include structures with suffixes

mathieulawnik opened the pull request:
In recursive calls of convert_struc() on suffixed include structures, the data holds the entire structure that the include structure is part of, so components of the include are suffixed, however the type is determined from the include structure alone, so the component list holds component names without their suffix. This leads to an ASSERTION_FAILURE error after component assignment fails because the names do not match.

mathieulawnik commented:
See: [](https://github.com/abapGit/abapGit/issues/6386)

mbtools commented on src/core/zcl_ajson.clas.locals_imp.abap:
```diff
@@ -1540,9 +1540,11 @@ class lcl_abap_to_json implementation.
 
       if <c>-as_include = abap_true.
 
+        ASSIGN COMPONENT <c>-name OF STRUCTURE iv_data TO FIELD-SYMBOL(<fs_data>).
```
This syntax is not available in 702. Move the definition of the field-symbol to the beginning of the method.

sbcgua commented:
Thanks for the contribution. Could you also please add a unit test covering the issue.

mathieulawnik commented:
Removing suffixes wasn't right. I've replaced the call to method get_components() with a call to get_included_view(). The structures are now converted flat instead of recursively.

mathieulawnik commented:
Is it possible to run the workflows locally ? The unit test workflow failed on my previous commit and I would like to check why.

mbtools commented:
The method is not implemented in open-abap which is used to run UTs. 
https://github.com/open-abap/open-abap-core/blob/main/src/rtti/cl_abap_structdescr.clas.abap

You could try to add it there or work with @larshp to get it implemented

sbcgua commented:
The code change looks good from my perspective. So can be merged as soon as the CI stuff passes.

larshp commented:
I'll fix the issue in open-abap, just need some time...

larshp commented:
its kind of a mess for me to fix, so it will be a while, suggest adding a skip to https://github.com/sbcgua/ajson/blob/master/transpile_for_testing.json#L22

mathieulawnik commented:
An alternative to using get_included_view() would be to pass the prefix to the convert_struc() method and append it to the node names.

larshp commented:
keep the code IMHO, avoid workarounds due to the open-abap runtime

sbcgua commented:
> its kind of a mess for me to fix, so it will be a while, suggest adding a skip to https://github.com/sbcgua/ajson/blob/master/transpile_for_testing.json#L22

:+1:

sbcgua commented:
Merging. I'll add the test exception. @mathieulawnik Thanks for the contribution!

sbcgua commented:
@larshp Did you fixed the issue on the transpiler side? Because the tests pass now. (If yes, thank you! =)

larshp commented:
hmm, no, I think its fixed in https://github.com/open-abap/open-abap-core/pull/737/commits/72b6a4d57bbce6ac121b320be585e7897f79ee71, but its not merged yet

mbtools commented:
looks like the fix was in https://github.com/open-abap/open-abap-core/pull/737
